partner_contact_department/tests/test_recursion.py

```diff
@@ -0,0 +1,26 @@
+# © 2016 Tecnativa - Vicent Cubells
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl-3.0).
+from odoo.exceptions import UserError
+from odoo.tests import common
+
+
+class TestRecursion(common.SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestRecursion, cls).setUpClass()
+        cls.department_obj = cls.env["res.partner.department"]
+
+        # Instances
+        cls.dpt1 = cls.department_obj.create({"name": "Dpt. 1"})
+        cls.dpt2 = cls.department_obj.create(
+            {"name": "Dep. 2", "parent_id": cls.dpt1.id}
+        )
+
+    def test_recursion(self):
+        """ Testing recursion """
+        self.dpt3 = self.department_obj.create(
+            {"name": "Dep. 3", "parent_id": self.dpt2.id}
+        )
+        # Creating a parent's child department using dpt1.
+        with self.assertRaises(UserError):
```

Comment:
Are you sure test is correct as in coverage result, the line with ValidationError is not reached ?

Reply:
![image](https://user-images.githubusercontent.com/19529533/81940409-db7bb600-95f7-11ea-9ac3-645b88a971cc.png)

Reply:
I think this is another error that is raised

Reply:
Hi @rousseldenis ,
we changed assertRaises to `UserError`, because with `ValidationError` the assert doesn't work properly.
In `_check_parent_id` function I think 'validationError' isn't correct. It should be  `UserError`, following the same logic.
If you have any better assertRaises we can change that.

Reply:
Nope. Constraints should always raise ValidationError.

What kind of problem did you have?

Reply:
Running tests, they returned a fatal error.
AssertRaises doesn't catch the 'ValidationError' type and it does catch the `UserError`.

Reply:
You didn't check the error correctly :smile: 

Just tested, this is the standard Odoo error that has been raised. You don't need the recursion check. Odoo does it for you!

https://github.com/odoo/odoo/blob/13.0/odoo/models.py#L3974

Reply:
If Odoo checks the recursion, `test_recursion.py` file doesn't need it, should we erase it?

Reply:
Yes, you can and the constraint too.

Reply:
Or if you want to be sure further versions check it, just remove the constraint and not the test. It's yours